<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; biokit  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/flavicon.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tutorials" href="../tutorials/index.html" />
    <link rel="prev" title="About" href="../about/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about/index.html">About</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#general-usage">General usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calling-functions">Calling functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#function-aliases">Function aliases</a></li>
<li class="toctree-l3"><a class="reference internal" href="#command-line-interfaces">Command line interfaces</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#alignment-based-functions">Alignment-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#alignment-length">Alignment length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-recoding">Alignment recoding</a></li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-summary">Alignment summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="#consensus-sequence">Consensus sequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#constant-sites">Constant sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parsimony-informative-sites">Parsimony informative sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="#position-specific-score-matrix">Position specific score matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#variable-sites">Variable sites</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#coding-sequence-based-functions">Coding sequence-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#gc-content-first-codon-position">GC content first codon position</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gc-content-second-codon-position">GC content second codon position</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gc-content-third-codon-position">GC content third codon position</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene-wise-relative-synonymous-codon-usage-grscu">Gene-wise relative synonymous codon usage (gRSCU)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#relative-synonymous-codon-usage">Relative synonymous codon usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#translate-sequence">Translate sequence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastq-file-functions">FASTQ file functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastq-read-lengths">FASTQ read lengths</a></li>
<li class="toctree-l3"><a class="reference internal" href="#subset-pe-fastq-reads">Subset PE FASTQ reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#subset-se-fastq-reads">Subset SE FASTQ reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-pe-adapters-fastq-reads">Trim PE adapters FASTQ reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-pe-fastq-reads">Trim PE FASTQ reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-se-adapters-fastq-reads">Trim SE adapters FASTQ reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-se-fastq-reads">Trim SE FASTQ reads</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#genome-functions">Genome functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#gc-content">GC content</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genome-assembly-metrics">Genome assembly metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#l50">L50</a></li>
<li class="toctree-l3"><a class="reference internal" href="#l90">L90</a></li>
<li class="toctree-l3"><a class="reference internal" href="#longest-scaffold">Longest scaffold</a></li>
<li class="toctree-l3"><a class="reference internal" href="#n50">N50</a></li>
<li class="toctree-l3"><a class="reference internal" href="#n90">N90</a></li>
<li class="toctree-l3"><a class="reference internal" href="#number-of-large-scaffolds">Number of large scaffolds</a></li>
<li class="toctree-l3"><a class="reference internal" href="#number-of-scaffolds">Number of scaffolds</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sum-of-scaffold-lengths">Sum of scaffold lengths</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-summary-and-processing-functions">Sequence summary and processing functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#character-frequency">Character frequency</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-fasta-entry-faidx">Get FASTA entry (faidx)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#file-format-converter">File format converter</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiple-line-to-single-line-fasta">Multiple line to single line FASTA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#remove-fasta-entry">Remove FASTA entry</a></li>
<li class="toctree-l3"><a class="reference internal" href="#remove-short-sequences">Remove short sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rename-fasta-entries">Rename FASTA entries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reorder-by-sequence-length">Reorder by sequence length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sequence-complement">Sequence complement</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sequence-length">Sequence length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#single-line-to-multiple-line-fasta">Single line to multiple line fasta</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../change_log/index.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../frequently_asked_questions/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">biokit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<span id="id1"></span><h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h1>
<p>BioKIT helps process and analyze data for bioinformatics research.</p>
<p>Generally, all functions are designed to help understand the contents of
various files including those representing alignments, coding sequences, fastq
files, and genomes.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="general-usage">
<h2>General usage<a class="headerlink" href="#general-usage" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="calling-functions">
<h3>Calling functions<a class="headerlink" href="#calling-functions" title="Permalink to this heading"></a></h3>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit &lt;command&gt; <span class="o">[</span>optional <span class="nb">command</span> arguments<span class="o">]</span>
</pre></div>
</div>
<p>Command specific help messages can be viewed by adding a
-h/--help argument after the command. For example, to see the
to see the help message for the command ‘alignment_summary’, execute:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit alignment_summary -h
<span class="c1"># or</span>
biokit alignment_summary --help
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="function-aliases">
<h3>Function aliases<a class="headerlink" href="#function-aliases" title="Permalink to this heading"></a></h3>
<p>Each function comes with aliases to save the user some
key strokes. For example, to get the help message for the ‘alignment_summary’
function, you can type:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit aln_summary -h
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="command-line-interfaces">
<h3>Command line interfaces<a class="headerlink" href="#command-line-interfaces" title="Permalink to this heading"></a></h3>
<p>All functions (including aliases) can also be executed using
a command-line interface that starts with <em>bk_</em>. For example, instead of typing
the previous command to get the help message of the alignment_summary function,
you can type:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>bk_alignment_summary -h
<span class="c1"># or</span>
bk_aln_summary -h
</pre></div>
</div>
<p>All possible function names are specified at the top of each function section.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="alignment-based-functions">
<h2>Alignment-based functions<a class="headerlink" href="#alignment-based-functions" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="alignment-length">
<h3>Alignment length<a class="headerlink" href="#alignment-length" title="Permalink to this heading"></a></h3>
<p>Function names: alignment_length; aln_len <br/>
Command line interface: bk_alignment_length; bk_aln_len</p>
<p>Calculate the length of an alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit alignment_length &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="alignment-recoding">
<h3>Alignment recoding<a class="headerlink" href="#alignment-recoding" title="Permalink to this heading"></a></h3>
<p>Function names: alignment_recoding; aln_recoding; recode <br/>
Command line interface: bk_alignment_recoding; bk_aln_recoding; bk_recode</p>
<p>Recode alignments using reduced character states.</p>
<p>Alignments can be recoded using established or
custom recoding schemes. Recoding schemes are
specified using the -c/–code argument. Custom
recoding schemes can be used and should be formatted
as a two column file wherein the first column is the
recoded character and the second column is the character
in the alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit alignment_recoding &lt;fasta&gt; <span class="o">[</span>-c/--code &lt;code&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Codes for which recoding scheme to use: <br/>
<strong>RY-nucleotide</strong> <br/>
R = purines (i.e., A and G) <br/>
Y = pyrimidines (i.e., T and C)</p>
<p><strong>Dayhoff-6</strong> <br/>
0 = A, G, P, S, and T <br/>
1 = D, E, N, and Q <br/>
2 = H, K, and R <br/>
3 = I, L, M, and V <br/>
4 = F, W, and Y <br/>
5 = C</p>
<p><strong>SandR-6</strong> <br/>
0 = A, P, S, and T <br/>
1 = D, E, N, and G <br/>
2 = Q, K, and R <br/>
3 = M, I, V, and L <br/>
4 = W and C <br/>
5 = F, Y, and H <br/></p>
<p><strong>KGB-6</strong> <br/>
0 = A, G, P, and S <br/>
1 = D, E, N, Q, H, K, R, and T <br/>
2 = M, I, and L <br/>
3 = W <br/>
4 = F and Y <br/>
5 = C and V <br/></p>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-c/--code</em>: argument to specify the recoding scheme to use</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="alignment-summary">
<h3>Alignment summary<a class="headerlink" href="#alignment-summary" title="Permalink to this heading"></a></h3>
<p>Function names: alignment_summary; aln_summary <br/>
Command line interface: bk_alignment_summary; bk_aln_summary</p>
<p>Summary statistics for an alignment. Reported
statistics include alignment length, number of taxa,
number of parsimony sites, number of variable sites,
number of constant sites, frequency of each character
(including gaps, which are considered to be ‘-’ or ‘?’).</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit alignment_summary &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="consensus-sequence">
<h3>Consensus sequence<a class="headerlink" href="#consensus-sequence" title="Permalink to this heading"></a></h3>
<p>Function names: consensus_sequence; con_seq <br/>
Command line interface: bk_consensus_sequence; bk_con_seq</p>
<p>Generates a consequence from a multiple sequence alignment file in FASTA format.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit consensus_sequence &lt;fasta&gt; -t/--threshold &lt;threshold&gt; -ac/--ambiguous_character &lt;ambiguous character&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be an alignment fasta file
<em>&lt;threshold&gt;</em>: threshold for how common a residue must be to be represented
<em>&lt;ambiguous character&gt;</em>: the ambiguity character to use. Default is ‘N’</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="constant-sites">
<h3>Constant sites<a class="headerlink" href="#constant-sites" title="Permalink to this heading"></a></h3>
<p>Function names: constant_sites; con_sites <br/>
Command line interface: bk_constant_sites; bk_con_sites</p>
<p>Calculate the number of constant sites in an alignment.</p>
<p>Constant sites are defined as a site in an
alignment with the same nucleotide or amino
acid sequence (excluding gaps) among all taxa.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit constant_sites &lt;fasta&gt; <span class="o">[</span>-v/<span class="se">\-\-</span>verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file
<em>-v/--verbose</em>: optional argument to print site-by-site categorization</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="parsimony-informative-sites">
<h3>Parsimony informative sites<a class="headerlink" href="#parsimony-informative-sites" title="Permalink to this heading"></a></h3>
<p>Function names: parsimony_informative_sites; pi_sites; pis <br/>
Command line interface: bk_parsimony_informative_sites; bk_pi_sites; bk_pis</p>
<p>Calculate the number of parsimony informative
sites in an alignment.</p>
<p>Parsimony informative sites are defined as a
site in an alignment with at least two nucleotides
or amino acids that occur at least twice.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit parsimony_informative_sites &lt;fasta&gt; <span class="o">[</span>-v/<span class="se">\-\-</span>verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-v/--verbose</em>: optional argument to print site-by-site categorization</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="position-specific-score-matrix">
<h3>Position specific score matrix<a class="headerlink" href="#position-specific-score-matrix" title="Permalink to this heading"></a></h3>
<p>Function names: position_specific_score_matrix; pssm <br/>
Command line interface: bk_position_specific_score_matrix; bk_pssm</p>
<p>Generates a position specific score matrix for an alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit position_specific_score_matrix &lt;fasta&gt; <span class="o">[</span>-ac/--ambiguous_character &lt;ambiguous character&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be an alignment fasta file
<em>&lt;ambiguous character&gt;</em>: the ambiguity character to use. Default is ‘N’</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="variable-sites">
<h3>Variable sites<a class="headerlink" href="#variable-sites" title="Permalink to this heading"></a></h3>
<p>Function names: variable_sites; var_sites; vs <br/>
Command line interface: bk_variable_sites; bk_var_sites; bk_vs</p>
<p>Calculate the number of variable sites in an
alignment.</p>
<p>Variable sites are defined as a site in an
alignment with at least two nucleotide or amino
acid characters among all taxa.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit variable_sites &lt;fasta&gt; <span class="o">[</span>-v/<span class="se">\-\-</span>verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file
<em>-v/--verbose</em>: optional argument to print site-by-site categorization</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="coding-sequence-based-functions">
<h2>Coding sequence-based functions<a class="headerlink" href="#coding-sequence-based-functions" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="gc-content-first-codon-position">
<h3>GC content first codon position<a class="headerlink" href="#gc-content-first-codon-position" title="Permalink to this heading"></a></h3>
<p>Function names: gc_content_first_position; gc1 <br/>
Command line interface: bk_gc_content_first_position; bk_gc1</p>
<p>Calculate GC content of the first codon position.
The input must be the coding sequence of a gene or
genes. All genes are assumed to have sequence lengths
divisible by three.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit gc_content_first_position &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/--verbose</em>: optional argument to print the GC content of each fasta entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="gc-content-second-codon-position">
<h3>GC content second codon position<a class="headerlink" href="#gc-content-second-codon-position" title="Permalink to this heading"></a></h3>
<p>Function names: gc_content_second_position; gc2 <br/>
Command line interface: bk_gc_content_second_position; bk_gc2</p>
<p>Calculate GC content of the second codon position.
The input must be the coding sequence of a gene or
genes. All genes are assumed to have sequence lengths
divisible by three.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit gc_content_second_position &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/--verbose</em>: optional argument to print the GC content of each fasta entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="gc-content-third-codon-position">
<h3>GC content third codon position<a class="headerlink" href="#gc-content-third-codon-position" title="Permalink to this heading"></a></h3>
<p>Function names: gc_content_third_position; gc3 <br/>
Command line interface: bk_gc_content_third_position; bk_gc3</p>
<p>Calculate GC content of the third codon position.
The input must be the coding sequence of a gene or
genes. All genes are assumed to have sequence lengths
divisible by three.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit gc_content_third_position &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/--verbose</em>: optional argument to print the GC content of each fasta entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="gene-wise-relative-synonymous-codon-usage-grscu">
<h3>Gene-wise relative synonymous codon usage (gRSCU)<a class="headerlink" href="#gene-wise-relative-synonymous-codon-usage-grscu" title="Permalink to this heading"></a></h3>
<p>Function names: gene_wise_relative_synonymous_codon_usage; gene_wise_rscu; gw_rscu; grscu <br/>
Command line interface: bk_gene_wise_relative_synonymous_codon_usage; bk_gene_wise_rscu; bk_gw_rscu; bk_grscu</p>
<p>Calculate gene-wise relative synonymous codon usage (gRSCU).</p>
<p>Codon usage bias examines biases for codon usage of
a particular gene. We adapted RSCU to be applied to
individual genes rather than only codons. Specifically,
gRSCU is the mean (or median) RSCU value observed
in a particular gene. This provides insight into how
codon usage bias influences codon usage for a particular
gene. This function also outputs the standard deviation
of RSCU values for a given gene.</p>
<p>The output is col 1: the gene identifier; col 2: the
gRSCU based on the mean RSCU value observed in a gene;
col 3: the gRSCU based on the median RSCU value observed
in a gene; and the col 4: the standard deviation of
RSCU values observed in a gene.</p>
<p>Custom genetic codes can be used as input and should
be formatted with the codon in first column and the
resulting amino acid in the second column.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit gene_wise_relative_synonymous_codon_usage &lt;fasta&gt; <span class="o">[</span>-tt/--translation_table &lt;code&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-tt/--translation_table</em>: optional argument of the code for the translation
table to be used. Default: 1, which is the standard code.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="relative-synonymous-codon-usage">
<h3>Relative synonymous codon usage<a class="headerlink" href="#relative-synonymous-codon-usage" title="Permalink to this heading"></a></h3>
<p>Function names: relative_synonymous_codon_usage; rscu <br/>
Command line interface: bk_relative_synonymous_codon_usage; bk_rscu</p>
<p>Calculate relative synonymous codon usage.</p>
<p>Relative synonymous codon usage is the ratio
of the observed frequency of codons over the
expected frequency given that all the synonymous
codons for the same amino acids are used equally.</p>
<p>Custom genetic codes can be used as input and should
be formatted with the codon in first column and the
resulting amino acid in the second column.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit relative_synonymous_codon_usage &lt;fasta&gt; <span class="o">[</span>-tt/--translation_table &lt;code&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-tt/--translation_table</em>: optional argument of the code for the translation
table to be used. Default: 1, which is the standard code.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="translate-sequence">
<h3>Translate sequence<a class="headerlink" href="#translate-sequence" title="Permalink to this heading"></a></h3>
<p>Function names: translate_sequence; translate_seq; trans_seq <br/>
Command line interface: bk_translate_sequence; bk_translate_seq; bk_trans_seq</p>
<p>Translates coding sequences to amino acid
sequences. Sequences can be translated using
diverse genetic codes. For codons that can
encode two amino acids (e.g., TAG encodes
Glu or STOP in the Blastocrithidia Nuclear Code),
the standard genetic code is used.</p>
<p>Custom genetic codes can be used as input and should
be formatted with the codon in first column and the
resulting amino acid in the second column.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit translate_sequence &lt;fasta&gt; <span class="o">[</span>-tt/--translation_table &lt;code&gt; -o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-tt/--translation_table</em>: optional argument of the code for the translation
table to be used. Default: 1, which is the standard code.
<em>-o/--output</em>: optional argument to write the translated fasta file to.
Default output has the same name as the input file with the suffix “.translated.fa” added
to it.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="fastq-file-functions">
<h2>FASTQ file functions<a class="headerlink" href="#fastq-file-functions" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="fastq-read-lengths">
<h3>FASTQ read lengths<a class="headerlink" href="#fastq-read-lengths" title="Permalink to this heading"></a></h3>
<p>Function names: fastq_read_lengths; fastq_read_lens <br/>
Command line interface: bk_fastq_read_lengths; bk_fastq_read_lens</p>
<p>Determine lengths of FASTQ reads.</p>
<p>Using default arguments, the average and
standard deviation of read lengths in a
FASTQ file will be reported. To obtain
the lengths of all FASTQ reads, use the
verbose option.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit fastq_read_lengths &lt;fasta&gt; <span class="o">[</span>-tt/--translation_table &lt;code&gt; -o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq&gt;</em>: first argument after function name should be a FASTQ file <br/>
<em>-v/--verbose</em>: print length of each FASTQ read</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="subset-pe-fastq-reads">
<h3>Subset PE FASTQ reads<a class="headerlink" href="#subset-pe-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: subset_pe_fastq_reads; subset_pe_fastq <br/>
Command line interface: bk_subset_pe_fastq_reads; bk_subset_pe_fastq</p>
<p>Subset paired-end FASTQ data.</p>
<p>Subsetting FASTQ data may be helpful for
running test scripts or achieving equal
coverage between samples. A percentage of
total reads in paired-end FASTQ data can
be obtained with this function. Random
subsamples are obtained using seeds for
reproducibility. If no seed is specified,
a seed is generated based off of the date
and time. During subsampling, paired-end
information is maintained.</p>
<p>Files are outputed with the suffix “_subset.fq”</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit subset_pe_fastq_reads &lt;fastq1&gt; &lt;fastq2&gt; <span class="o">[</span>-p/--percent &lt;percent&gt; -s/--seed &lt;seed&gt; -o/--output_file &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq1&gt;</em>: first argument after function name should be the name of one of the fastq files <br/>
<em>&lt;fastq2&gt;</em>: first argument after function name should be the name of the other fastq file <br/>
<em>-p/--percent</em>: percentage of reads to maintain in subsetted data <br/>
<em>-s/--seed</em>: seed for random sampling</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="subset-se-fastq-reads">
<h3>Subset SE FASTQ reads<a class="headerlink" href="#subset-se-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: subset_se_fastq_reads; subset_se_fastq <br/>
Command line interface: bk_subset_se_fastq_reads; bk_subset_se_fastq</p>
<p>Subset single-end FASTQ data.</p>
<p>Subsetting FASTQ data may be helpful for
running test scripts or achieving equal
coverage between samples. A percentage of
total reads in single-end FASTQ data can
be obtained with this function. Random
subsamples are obtained using seeds for
reproducibility. If no seed is specified,
a seed is generated based off of the date
and time.</p>
<p>Output files will have the suffix “_subset.fq”</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit subset_se_fastq_reads &lt;fastq&gt; <span class="o">[</span>-p/--percent &lt;percent&gt; -s/--seed &lt;seed&gt; -o/--output_file &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq&gt;</em>: first argument after function name should be the name of one of the fastq files <br/>
<em>-p/--percent</em>: percentage of reads to maintain in subsetted data <br/>
<em>-s/--seed</em>: seed for random sampling <br/>
<em>-o/--output_file</em>: specify the name of the output file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="trim-pe-adapters-fastq-reads">
<h3>Trim PE adapters FASTQ reads<a class="headerlink" href="#trim-pe-adapters-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: trim_pe_adapters_fastq_reads; trim_pe_adapters_fastq <br/>
Command line interface: bk_trim_pe_adapters_fastq_reads; bk_trim_pe_adapters_fastq</p>
<p>Trim adapters from paired-end FastQ data.</p>
<p>FASTQ data will be trimmed according to
exact match to known adapter sequences.</p>
<p>Output file has the suffix “_adapter_removed.fq”
or can be named by the user with the
output_file argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit trim_pe_adapters_fastq_reads &lt;fastq1&gt; &lt;fastq2&gt; <span class="o">[</span>-a/--adapters TruSeq2-PE -l/--length <span class="m">20</span><span class="o">]</span>
</pre></div>
</div>
<p>Adapaters available: <br/>
NexteraPE-PE <br/>
TruSeq2-PE <br/>
TruSeq2-SE <br/>
TruSeq3-PE-2 <br/>
TruSeq3-PE <br/>
TruSeq3-SE</p>
<p>Options: <br/>
<em>&lt;fastq1&gt;</em>: first argument after function name should be the name of one of the fastq files <br/>
<em>&lt;fastq2&gt;</em>: first argument after function name should be the name of the other fastq file <br/>
<em>-a/--adapters</em>: adapter sequences to trim. Default: TruSeq2-PE <br/>
<em>-l/--length</em>: minimum length of read to be kept. Default: 20</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="trim-pe-fastq-reads">
<h3>Trim PE FASTQ reads<a class="headerlink" href="#trim-pe-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: trim_pe_fastq_reads; trim_pe_fastq <br/>
Command line interface: bk_trim_pe_fastq_reads; bk_trim_pe_fastq</p>
<p>Quality trim paired-end FastQ data.</p>
<p>FASTQ data will be trimmed according to
quality score and length of the reads.
Users can specify quality and length
thresholds. Paired reads that are
maintained and saved to files with the
suffix “_paired_trimmed.fq.” Single
reads that passed quality thresholds are
saved to files with the suffix
“_unpaired_trimmed.fq.”</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit trim_pe_fastq_reads &lt;fastq1&gt; &lt;fastq2&gt; <span class="o">[</span>-m/--minimum <span class="m">20</span> -l/--length <span class="m">20</span><span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq1&gt;</em>: first argument after function name should be the name of one of the fastq files <br/>
<em>&lt;fastq2&gt;</em>: first argument after function name should be the name of the other fastq file <br/>
<em>-m/--minimum</em>: minimum quality of read to be kept. Default: 20 <br/>
<em>-l/--length</em>: minimum length of read to be kept. Default: 20</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="trim-se-adapters-fastq-reads">
<h3>Trim SE adapters FASTQ reads<a class="headerlink" href="#trim-se-adapters-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: trim_se_adapters_fastq_reads; trim_se_adapters_fastq <br/>
Command line interface: bk_trim_se_adapters_fastq_reads; bk_trim_se_adapters_fastq</p>
<p>Trim adapters from single-end FastQ data.</p>
<p>FASTQ data will be trimmed according to
exact match to known adapter sequences.</p>
<p>Output file has the suffix “_adapter_removed.fq”
or can be named by the user with the
output_file argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit trim_se_adapters_fastq_reads &lt;fastq&gt; <span class="o">[</span>-a/--adapters TruSeq2-SE -l/--length <span class="m">20</span><span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq&gt;</em>: first argument after function name should be the fastq file <br/>
<em>-a/--adapters</em>: adapter sequences to trim. Default: TruSeq2-SE <br/>
<em>-l/--length</em>: minimum length of read to be kept. Default: 20 <br/>
<em>-o/--output_file</em>: name of the output file of trimmed reads</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="trim-se-fastq-reads">
<h3>Trim SE FASTQ reads<a class="headerlink" href="#trim-se-fastq-reads" title="Permalink to this heading"></a></h3>
<p>Function names: trim_se_fastq_reads; trim_se_fastq <br/>
Command line interface: bk_trim_se_fastq_reads; bk_trim_se_fastq</p>
<p>Quality trim single-end FastQ data.</p>
<p>FASTQ data will be trimmed according to
quality score and length of the reads.
Users can specify quality and length
thresholds. Output file has the suffix
“_trimmed.fq” or can be named by the user
with the output_file argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit trim_se_fastq_reads &lt;fastq&gt; <span class="o">[</span>-m/--minimum <span class="m">20</span> -l/--length <span class="m">20</span><span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fastq&gt;</em>: first argument after function name should be the fastq file <br/>
<em>-m/--minimum</em>: minimum quality of read to be kept. Default: 20 <br/>
<em>-l/--length</em>: minimum length of read to be kept. Default: 20 <br/>
<em>-o/--output_file</em>: name of the output file of trimmed reads</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="genome-functions">
<h2>Genome functions<a class="headerlink" href="#genome-functions" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="gc-content">
<h3>GC content<a class="headerlink" href="#gc-content" title="Permalink to this heading"></a></h3>
<p>Function names: gc_content; gc <br/>
Command line interface: bk_gc_content; bk_gc</p>
<p>Calculate GC content of a fasta file.</p>
<p>GC content is the fraction of bases that are
either guanines or cytosines. To obtain
GC content per FASTA entry, use the verbose
option.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit gc_content &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/--verbose</em>: optional argument to print the GC content of each fasta entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="genome-assembly-metrics">
<h3>Genome assembly metrics<a class="headerlink" href="#genome-assembly-metrics" title="Permalink to this heading"></a></h3>
<p>Function names: genome_assembly_metrics; assembly_metrics <br/>
Command line interface: bk_genome_assembly_metrics; bk_assembly_metrics</p>
<p>Calculate L50, L90, N50, N90, GC content, assembly size,
number of scaffolds, number and sum length
of large scaffolds, frequency of A, T, C, and G.</p>
<p>L50: The smallest number of contigs whose length sum makes up half of the genome size. <br/>
L90: The smallest number of contigs whose length sum makes up 90% of the genome size. <br/>
N50: The sequence length of the shortest contig at half of the genome size. <br/>
N90: The sequence length of the shortest contig at 90% of the genome size. <br/>
GC content: The fraction of bases that are either guanines or cytosines. <br/>
Assembly size: The sum length of all contigs in an assembly. <br/>
Number of scaffolds: The total number of scaffolds in an assembly. <br/>
Number of large scaffolds: The total number of scaffolds that are greater than the threshold for small scaffolds. <br/>
Sum length of large scaffolds: The sum length of all large scaffolds. <br/>
Frequency of A: The number of occurences of A corrected by assembly size. <br/>
Frequency of T: The number of occurences of T corrected by assembly size. <br/>
Frequency of C: The number of occurences of C corrected by assembly size. <br/>
Frequency of G: The number of occurences of G corrected by assembly size.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit genome_assembly_metrics &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-t/--threshold</em>: threshold for what is considered a large scaffold.
Only scaffolds with a length greater than this value will be counted. Default: 500</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="l50">
<h3>L50<a class="headerlink" href="#l50" title="Permalink to this heading"></a></h3>
<p>Function names: l50 <br/>
Command line interface: bk_l50</p>
<p>Calculates L50 for a genome assembly.</p>
<p>L50 is the smallest number of contigs whose length sum
makes up half of the genome size.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit l50 &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="l90">
<h3>L90<a class="headerlink" href="#l90" title="Permalink to this heading"></a></h3>
<p>Function names: l90 <br/>
Command line interface: bk_l90</p>
<p>Calculates L90 for a genome assembly.</p>
<p>L90 is the smallest number of contigs whose length sum
makes up 90% of the genome size.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit l90 &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="longest-scaffold">
<h3>Longest scaffold<a class="headerlink" href="#longest-scaffold" title="Permalink to this heading"></a></h3>
<p>Function names: longest_scaffold; longest_scaff; longest_contig; longest_cont <br/>
Command line interface: bk_longest_scaffold; bk_longest_scaff; bk_longest_contig; bk_longest_cont</p>
<p>Determine the length of the longest scaffold in a genome assembly.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit longest_scaffold &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="n50">
<h3>N50<a class="headerlink" href="#n50" title="Permalink to this heading"></a></h3>
<p>Function names: n50 <br/>
Command line interface: bk_n50</p>
<p>Calculates N50 for a genome assembly.</p>
<p>N50 is the sequence length of the shortest contig at 50% of the genome size.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit n50 &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="n90">
<h3>N90<a class="headerlink" href="#n90" title="Permalink to this heading"></a></h3>
<p>Function names: n90 <br/>
Command line interface: bk_n90</p>
<p>Calculates N90 for a genome assembly.</p>
<p>N90 is the sequence length of the shortest contig at 90% of the genome size.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit n90 &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="number-of-large-scaffolds">
<h3>Number of large scaffolds<a class="headerlink" href="#number-of-large-scaffolds" title="Permalink to this heading"></a></h3>
<p>Function names: number_of_large_scaffolds; num_of_lrg_scaffolds; number_of_large_contigs; num_of_lrg_cont <br/>
Command line interface: bk_number_of_large_scaffolds; bk_num_of_lrg_scaffolds; bk_number_of_large_contigs; bk_num_of_lrg_cont</p>
<p>Calculate number and total sequence length of
large scaffolds. Each value is represented as
column 1 and column 2 in the output, respectively.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit number_of_large_scaffolds &lt;fasta&gt; <span class="o">[</span>-t/--threshold &lt;int&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file
<em>-t/--threshold</em>: threshold for what is considered
a large scaffold. Only scaffolds with a length greater than this
value will be counted. Default: 500</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="number-of-scaffolds">
<h3>Number of scaffolds<a class="headerlink" href="#number-of-scaffolds" title="Permalink to this heading"></a></h3>
<p>Function names: number_of_scaffolds; num_of_scaffolds; number_of_contigs; num_of_cont <br/>
Command line interface: bk_number_of_scaffolds; bk_num_of_scaffolds; bk_number_of_contigs; bk_num_of_cont</p>
<p>Calculate the number of scaffolds or entries
in a FASTA file. In this way, a user can also
determine the number of predicted genes in a
coding sequence or protein FASTA file with this
function.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit number_of_scaffolds &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="sum-of-scaffold-lengths">
<h3>Sum of scaffold lengths<a class="headerlink" href="#sum-of-scaffold-lengths" title="Permalink to this heading"></a></h3>
<p>Function names: sum_of_scaffold_lengths; sum_of_contig_lengths <br/>
Command line interface: bk_sum_of_scaffold_lengths; bk_sum_of_contig_lengths</p>
<p>Determine the sum of scaffold lengths.</p>
<p>The intended use of this function is to determine
the length of a genome assembly, but can also be
used, for example, to determine the sum length
of all coding sequences.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit sum_of_scaffold_lengths &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="sequence-summary-and-processing-functions">
<h2>Sequence summary and processing functions<a class="headerlink" href="#sequence-summary-and-processing-functions" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<section id="character-frequency">
<h3>Character frequency<a class="headerlink" href="#character-frequency" title="Permalink to this heading"></a></h3>
<p>Function names: character_frequency; char_freq <br/>
Command line interface: bk_character_frequency; bk_char_freq</p>
<p>Calculate the frequency of characters in a FASTA file.</p>
<p>This can be used to determine the frequency of A, T, C, and G
in a genome or the frequency of amino acids in a proteome.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit character_frequency &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="get-fasta-entry-faidx">
<h3>Get FASTA entry (faidx)<a class="headerlink" href="#get-fasta-entry-faidx" title="Permalink to this heading"></a></h3>
<p>Function names: faidx; get_entry; ge <br/>
Command line interface: bk_faidx; bk_get_entry; bk_ge</p>
<p>Extracts sequence entry from fasta file.</p>
<p>This function works similarly to the faidx function
in samtools, but does not requiring an indexing the
sequence file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit faidx &lt;fasta&gt; <span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-e/--entry</em>: entry name to be extracted from the inputted fasta file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="file-format-converter">
<h3>File format converter<a class="headerlink" href="#file-format-converter" title="Permalink to this heading"></a></h3>
<p>Function names: file_format_converter; format_converter; ffc <br/>
Command line interface: bk_file_format_converter; bk_format_converter; bk_ffc</p>
<p>Converts a multiple sequence file from one format to another.</p>
<p>Acceptable file formats include FASTA, Clustal, MAF, Mauve,
Phylip, Phylip-sequential, Phylip-relaxed, and Stockholm.
Input and output file formats are specified with the
--input_file_format and --output_file_format arguments;
input and output files are specified with the --input_file
and --output_file arguments.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit file_format_converter -i/--input_file &lt;input_file&gt; -iff/--input_file_format &lt;input_file_format&gt;  -o/--output_file &lt;output_file&gt; -off/--output_file_format &lt;output_file_format&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>-i/--input_file</em>: input file name
<em>-iff/--input_file_format</em>: input file format
<em>-o/--output_file</em>: output file name
<em>-off/--output_file_format</em>: output file format</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="multiple-line-to-single-line-fasta">
<h3>Multiple line to single line FASTA<a class="headerlink" href="#multiple-line-to-single-line-fasta" title="Permalink to this heading"></a></h3>
<p>Function names: multiple_line_to_single_line_fasta; ml2sl <br/>
Command line interface: bk_multiple_line_to_single_line_fasta; bk_ml2sl</p>
<p>Converts FASTA files with multiple lines
per sequence to a FASTA file with the sequence
represented on one line.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit multiple_line_to_single_line_fasta &lt;fasta&gt; <span class="o">[</span>-o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-o/--output</em>: optional argument to name the output file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="remove-fasta-entry">
<h3>Remove FASTA entry<a class="headerlink" href="#remove-fasta-entry" title="Permalink to this heading"></a></h3>
<p>Function names: remove_fasta_entry <br/>
Command line interface: bk_remove_fasta_entry</p>
<p>Remove FASTA entry from multi-FASTA file.</p>
<p>Output will have the suffix “pruned.fa” unless
the user specifies a different output file name.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit remove_fasta_entry &lt;fasta&gt; -e/--entry &lt;entry&gt; <span class="o">[</span>-o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-e/--entry</em>: entry name to be removed from the inputted fasta file <br/>
<em>-o/--output</em>: optional argument to write the renamed fasta file to.
Default output has the same name as the input file with the suffix “pruned.fa”
added to it.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="remove-short-sequences">
<h3>Remove short sequences<a class="headerlink" href="#remove-short-sequences" title="Permalink to this heading"></a></h3>
<p>Function names: remove_short_sequences; remove_short_seqs <br/>
Command line interface: bk_remove_short_sequences; bk_remove_short_seqs</p>
<p>Remove short sequences from a multi-FASTA file.</p>
<p>Short sequences are defined as having a length
less than 500. Users can specify their own threshold.
All sequences greater than the threshold will be
kept in the resulting file.</p>
<p>Output will have the suffix “long_seqs.fa” unless
the user specifies a different output file name.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit remove_short_sequences &lt;fasta&gt; -t/--threshold &lt;threshold&gt; <span class="o">[</span>-o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-t/--threshold</em>: threshold for short sequences. Sequences greater
than this value will be kept <br/>
<em>-o/--output</em>: optional argument to write the renamed fasta file to.
Default output has the same name as the input file with the suffix “long_seqs.fa”
added to it.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="rename-fasta-entries">
<h3>Rename FASTA entries<a class="headerlink" href="#rename-fasta-entries" title="Permalink to this heading"></a></h3>
<p>Function names: rename_fasta_entries; rename_fasta <br/>
Command line interface: bk_rename_fasta_entries; bk_rename_fasta</p>
<p>Renames fasta entries.</p>
<p>Renaming fasta entries will follow the scheme of a tab-delimited
file wherein the first column is the current fasta entry name and
the second column is the new fasta entry name in the resulting
output alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>rename_fasta_entries &lt;fasta&gt; -i/--idmap &lt;idmap&gt; <span class="o">[</span>-o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-i/--idmap</em>: identifier map of current FASTA names (col1) and desired FASTA names (col2) <br/>
<em>-o/--output</em>: optional argument to name the output file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="reorder-by-sequence-length">
<h3>Reorder by sequence length<a class="headerlink" href="#reorder-by-sequence-length" title="Permalink to this heading"></a></h3>
<p>Function names: reorder_by_sequence_length; reorder_by_seq_len <br/>
Command line interface: bk_reorder_by_sequence_length; bk_reorder_by_seq_len</p>
<p>Reorder FASTA file entries from the longest entry
to the shortest entry.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit reorder_by_sequence_length &lt;fasta&gt; <span class="o">[</span>-o/--output &lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-o/--output</em>: optional argument to write the reordered fasta file to.
Default output has the same name as the input file with the suffix
“.reordered.fa” added to it.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="sequence-complement">
<h3>Sequence complement<a class="headerlink" href="#sequence-complement" title="Permalink to this heading"></a></h3>
<p>Function names: sequence_complement; seq_comp <br/>
Command line interface: bk_sequence_complement; bk_seq_comp</p>
<p>Generates the sequence complement for all entries
in a multi-FASTA file. To generate a reverse sequence
complement, add the -r/–reverse argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit sequence_complement &lt;fasta&gt; <span class="o">[</span>-r/--reverse<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-r/--reverse</em>: if used, the reverse complement sequence will be generated</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="sequence-length">
<h3>Sequence length<a class="headerlink" href="#sequence-length" title="Permalink to this heading"></a></h3>
<p>Function names: sequence_length; seq_len <br/>
Command line interface: bk_sequence_length; bk_seq_len</p>
<p>Calculate sequence length of each FASTA entry.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit sequence_length &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="single-line-to-multiple-line-fasta">
<h3>Single line to multiple line fasta<a class="headerlink" href="#single-line-to-multiple-line-fasta" title="Permalink to this heading"></a></h3>
<p>Function names: single_line_to_multiple_line_fasta; sl2ml <br/>
Command line interface: bk_single_line_to_multiple_line_fasta; bk_sl2ml</p>
<p>Calculate sequence length of each FASTA entry.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>biokit single_line_to_multiple_line_fasta &lt;fasta&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../about/index.html" class="btn btn-neutral float-left" title="About" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../tutorials/index.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022 Jacob L. Steenwyk.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-104875636-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-104875636-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>